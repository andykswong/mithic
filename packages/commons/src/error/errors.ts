import {
  ABORT_ERR, ERR_OPERATION_FAILED, INVALID_STATE_ERR, NOT_FOUND_ERR, NOT_SUPPORTED_ERR, TIMEOUT_ERR
} from './code.js';
import { CodedError, ErrorCodeDetailOptions, ErrorDetailOptions } from './error.js';

/** An AbortError. */
export class AbortError<T = unknown, E = unknown> extends CodedError<T, E> {
  public constructor(
    message = AbortError.name,
    options?: ErrorDetailOptions<T>,
  ) {
    super(message, { name: AbortError.name, code: ABORT_ERR, ...options });
  }
}

/** A TimeoutError. */
export class TimeoutError<T = unknown, E = unknown> extends CodedError<T, E> {
  public constructor(
    message = TimeoutError.name,
    options?: ErrorDetailOptions<T>,
  ) {
    super(message, { name: TimeoutError.name, code: TIMEOUT_ERR, ...options });
  }
}

/** A NotFoundError. */
export class NotFoundError<T = unknown, E = unknown> extends CodedError<T, E> {
  public constructor(
    message = NotFoundError.name,
    options?: ErrorDetailOptions<T>,
  ) {
    super(message, { name: NotFoundError.name, code: NOT_FOUND_ERR, ...options });
  }
}

/** A NotSupportedError. */
export class NotSupportedError<T = unknown, E = unknown> extends CodedError<T, E> {
  public constructor(
    message = NotSupportedError.name,
    options?: ErrorDetailOptions<T>,
  ) {
    super(message, { name: NotSupportedError.name, code: NOT_SUPPORTED_ERR, ...options });
  }
}

/** An InvalidStateError. */
export class InvalidStateError<T = unknown, E = unknown> extends CodedError<T, E> {
  public constructor(
    message = InvalidStateError.name,
    options?: ErrorDetailOptions<T>,
  ) {
    super(message, { name: InvalidStateError.name, code: INVALID_STATE_ERR, ...options });
  }
}

/** An OperationError. */
export class OperationError<T = unknown, E = unknown> extends CodedError<T, E> {
  public constructor(
    message = InvalidStateError.name,
    options?: ErrorCodeDetailOptions<T>,
  ) {
    super(message, { name: OperationError.name, code: ERR_OPERATION_FAILED, ...options });
  }
}
