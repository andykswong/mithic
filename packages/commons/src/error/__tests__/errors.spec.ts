import { describe, expect, it } from '@jest/globals';
import { AbortError, InvalidStateError, NotFoundError, NotSupportedError, OperationError, TimeoutError } from '../errors.ts';

describe(AbortError.name, () => {
  it('should initialize with given reason', () => {
    const detail = { 'detail': 'issue' };
    const cause = new Error('of cause');
    const error = new AbortError('test reason', { detail, cause });

    expect(error).toMatchSnapshot();
    expect({ ...error }).toMatchSnapshot();
    expect(error.cause).toBe(cause);
  });
});

describe(TimeoutError.name, () => {
  it('should initialize with given reason', () => {
    const detail = { 'detail': 'issue' };
    const cause = new Error('of cause');
    const error = new TimeoutError('test reason', { detail, cause });

    expect(error).toMatchSnapshot();
    expect({ ...error }).toMatchSnapshot();
    expect(error.cause).toBe(cause);
  });
});

describe(NotFoundError.name, () => {
  it('should initialize with given reason', () => {
    const detail = { 'detail': 'issue' };
    const cause = new Error('of cause');
    const error = new NotFoundError('test reason', { detail, cause });

    expect(error).toMatchSnapshot();
    expect({ ...error }).toMatchSnapshot();
    expect(error.cause).toBe(cause);
  });
});

describe(NotSupportedError.name, () => {
  it('should initialize with given reason', () => {
    const detail = { 'detail': 'issue' };
    const cause = new Error('of cause');
    const error = new NotSupportedError('test reason', { detail, cause });

    expect(error).toMatchSnapshot();
    expect({ ...error }).toMatchSnapshot();
    expect(error.cause).toBe(cause);
  });
});

describe(InvalidStateError.name, () => {
  it('should initialize with given reason', () => {
    const detail = { 'detail': 'issue' };
    const cause = new Error('of cause');
    const error = new InvalidStateError('test reason', { detail, cause });

    expect(error).toMatchSnapshot();
    expect({ ...error }).toMatchSnapshot();
    expect(error.cause).toBe(cause);
  });
});

describe(OperationError.name, () => {
  it('should initialize with given reason', () => {
    const detail = { 'detail': 'issue' };
    const cause = new Error('of cause');
    const error = new OperationError('test reason', { detail, cause });

    expect(error).toMatchSnapshot();
    expect({ ...error }).toMatchSnapshot();
    expect(error.cause).toBe(cause);
  });

  it('should initialize with given reason and code', () => {
    const detail = { 'detail': 'issue' };
    const cause = new Error('of cause');
    const error = new OperationError('test reason', { detail, cause, code: 'TEST_ERR' });

    expect(error).toMatchSnapshot();
    expect({ ...error }).toMatchSnapshot();
    expect(error.cause).toBe(cause);
  });
});
