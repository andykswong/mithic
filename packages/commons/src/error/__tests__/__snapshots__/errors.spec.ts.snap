// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbortError should initialize with given reason 1`] = `[AbortError: test reason]`;

exports[`AbortError should initialize with given reason 2`] = `
{
  "code": "ABORT_ERR",
  "detail": {
    "detail": "issue",
  },
  "name": "AbortError",
}
`;

exports[`InvalidStateError should initialize with given reason 1`] = `[InvalidStateError: test reason]`;

exports[`InvalidStateError should initialize with given reason 2`] = `
{
  "code": "INVALID_STATE_ERR",
  "detail": {
    "detail": "issue",
  },
  "name": "InvalidStateError",
}
`;

exports[`NotFoundError should initialize with given reason 1`] = `[NotFoundError: test reason]`;

exports[`NotFoundError should initialize with given reason 2`] = `
{
  "code": "NOT_FOUND_ERR",
  "detail": {
    "detail": "issue",
  },
  "name": "NotFoundError",
}
`;

exports[`NotSupportedError should initialize with given reason 1`] = `[NotSupportedError: test reason]`;

exports[`NotSupportedError should initialize with given reason 2`] = `
{
  "code": "NOT_SUPPORTED_ERR",
  "detail": {
    "detail": "issue",
  },
  "name": "NotSupportedError",
}
`;

exports[`OperationError should initialize with given reason 1`] = `[OperationError: test reason]`;

exports[`OperationError should initialize with given reason 2`] = `
{
  "code": "ERR_OPERATION_FAILED",
  "detail": {
    "detail": "issue",
  },
  "name": "OperationError",
}
`;

exports[`OperationError should initialize with given reason and code 1`] = `[OperationError: test reason]`;

exports[`OperationError should initialize with given reason and code 2`] = `
{
  "code": "TEST_ERR",
  "detail": {
    "detail": "issue",
  },
  "name": "OperationError",
}
`;

exports[`TimeoutError should initialize with given reason 1`] = `[TimeoutError: test reason]`;

exports[`TimeoutError should initialize with given reason 2`] = `
{
  "code": "TIMEOUT_ERR",
  "detail": {
    "detail": "issue",
  },
  "name": "TimeoutError",
}
`;
